File: blocking_vs_nonblocking/Demo2.js
const startTime = new Date();

console.log(`We are now on ${process.pid}`);
console.log("Starting non-blocking operation...");
const futureResult = performNonblockingOperation();

futureResult.then((res) => {
    console.log(res);
});

const endTime = new Date();
console.log(`Time taken: ${endTime - startTime} ms`);

async function performNonblockingOperation() {
    return new Promise(async (resolve, reject) => {
        console.log(`performing non-blocking operation on ${process.pid}`);
        try {
            await sleep(3000);
        } catch (error) {
            reject(error);
        }

        resolve('non-blocking operation completed');

    });
}

async function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

File: blocking_vs_nonblocking/Demo2.java
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CountDownLatch;

class Demo2 {
    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        CountDownLatch latch = new CountDownLatch(1);

        // Start non-blocking operation
        System.out.println("We are now on " + Thread.currentThread().getName());
        System.out.println("Starting non-blocking operation...");
        CompletableFuture<String> futureResult = performNonBlockingOperation();

        // this is to handle when the non-blocking operation finishes
        futureResult.whenComplete((res, t) -> {
            System.out.println(res);
        });

        long endTime = System.currentTimeMillis();
        System.out.println("Time taken: " + (endTime - startTime) + " ms");

        // this tells the program not to terminate before the futureResult resolves
        try {
			latch.await();
		} catch (InterruptedException e) {
			e.printStackTrace();
		} // Wait until the countdown reaches 0
    }

    private static CompletableFuture<String> performNonBlockingOperation() {
        return CompletableFuture.supplyAsync(() -> {
            System.out.println("performing non-blocking operation on " + Thread.currentThread().getName());
            try {
                // Simulate time-consuming I/O operation
                Thread.sleep(3000); // 3 seconds
            } catch (InterruptedException e) {
                e.printStackTrace();
            }

            return "non-blocking operation completed";
        });
    }
}
